### DISCUSSION TITLE
FastAPI is not working like expected in Async Mode

### DISCUSSION THREAD

**Discussion by shreyasY2k** [fastapi/fastapi#11261] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, File, UploadFile
import asyncio
import requests
from pdf2image import convert_from_bytes
import io
from app.logger import logger

async def split_pdf_to_images(pdf):
    try:
        pdf_bytes = await pdf.read()
        images = convert_from_bytes(pdf_bytes)
        image_bytes = [io.BytesIO(image.tobytes()) for image in images]
        return image_bytes
    except Exception as e:
        logger.error(f"Error occurred while splitting PDF: {str(e)}")
        raise e

async def call_api(data, api_url, data_field):
    try:
        response = requests.get(api_url)
        response.raise_for_status()
        return response.json()
    except Exception as e:
        logger.error(f"Error occurred while calling API {api_url}: {str(e)}")
        raise e
    
async def process_image(image):
    api_urls = ["https://url1/89731d00-f0fc-42c7-904a-1d2ccd383948", "https://url2/89731d00-f0fc-42c7-904a-1d2ccd383948", "https://url3/89731d00-f0fc-42c7-904a-1d2ccd383948"]
    response = await call_api(image, api_urls[0], "image")
    label = await call_api(response, api_urls[1], "data")
    if label["label"] == "KYC":
        response = await call_api(response, api_urls[2], "data")
    return response

app = FastAPI()

@app.post("/pdf-to-images")
async def pdf_to_images(pdf: UploadFile = File(...)):
    try:
        logger.info(f"Received PDF file: {pdf.filename}")
        images = await split_pdf_to_images(pdf)
        logger.info(f"PDF split into {len(images)} images")
        tasks = [asyncio.create_task(process_image(image)) for image in images]
        consolidated_response = await asyncio.gather(*tasks)
        logger.info("All API calls completed")
        logger.info(consolidated_response)
        return {"message": "PDF processed successfully", "data": consolidated_response}
    except Exception as e:
        logger.error(f"Error occurred: {str(e)}")
        return {"message": "Error occurred while processing the PDF"}
```

### Description

I need to build an API using FastAPI with the following requirements:

Should be Async

Accepts a PDF and input

Split the pdf to images

call 3 api endpoints for each image synchronously

It should be something like promise.all of nodejs where you wait for all the apis calls of all the images are completed

When I do so in FastAPI the requests are not accepted/processed parallely, while in node it is otherwise.

** Note: I'm using single uvicorn worker and as per FastAPI calim single worker should be able to process parllel requests.**

Please help me and let me know what am I doing wrong.

Thanks in advance.

### Operating System

Linux

### Operating System Details

Ubuntu 20.04.6

### FastAPI Version

0.78.0

### Pydantic Version

1.9.1

### Python Version

3.7

### Additional Context

Refer to the screenshot below.

![Log_Comparision](https://github.com/tiangolo/fastapi/assets/48048098/b76f48b8-2c8d-43a0-a650-a20920308fdb)

LHS is the FastAPI log and RHS is NodeJS log file

**Comment by SarloAkrobata** [comment, marked as answer]
It has nothing to do with fastapi. `response = requests.get(api_url)` is blocker, this is sync task. You need to use some of async libs like httpx ot aiohttp

**Reply by shreyasY2k** [reply]
Thanks! Got it working using httpx!
